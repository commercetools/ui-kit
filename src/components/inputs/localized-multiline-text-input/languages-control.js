import React from 'react';
import PropTypes from 'prop-types';
import { useIntl } from 'react-intl';
import FlatButton from '@commercetools-uikit/flat-button';
import { WorldIcon } from '@commercetools-uikit/icons';
import messages from '../../internals/messages/localized-input';

const LanguagesControl = (props) => {
  const intl = useIntl();
  return (
    <FlatButton
      icon={<WorldIcon />}
      label={intl.formatMessage(
        props.isClosed ? messages.show : messages.hide,
        {
          remainingLanguages: props.remainingLanguages,
        }
      )}
      onClick={props.onClick}
      isDisabled={props.isDisabled}
    />
  );
};

LanguagesControl.displayName = 'LanguagesControl';
LanguagesControl.propTypes = {
  onClick: PropTypes.func.isRequired,
  isClosed: PropTypes.bool,
  isDisabled: PropTypes.bool,
  remainingLanguages: PropTypes.number.isRequired,
};

export default LanguagesControl;
