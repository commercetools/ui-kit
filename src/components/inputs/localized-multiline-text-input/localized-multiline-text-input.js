import React from 'react';
import PropTypes from 'prop-types';
import requiredIf from 'react-required-if';
import mapValues from 'lodash.mapvalues';
import oneLine from 'common-tags/lib/oneLine';
import { injectIntl } from 'react-intl';
import Spacings from '../../spacings';
import Constraints from '../../constraints';
import {
  sortLanguages,
  createLocalizedDataAttributes,
  getHasErrorOnRemainingLanguages,
  getHasWarningOnRemainingLanguages,
  isTouched,
  omitEmptyTranslations,
  isEmpty,
  createLocalizedString,
  getId,
  getName,
} from '../../../utils/localized';
import LanguagesControl from './languages-control';
import TranslationInput from './translation-input';
import RequiredValueErrorMessage from './required-value-error-message';

// This component supports expanding/collapsing multiline inputs, but it also
// supports showing/hiding the remaining languages.
// These two features are both about opening/closing something, and so the code
// can get quite confusing. We try to stick to expand/collapse for the
// multiline inputs, while we use show/hide/open/close for the remaining
// languages.
export class LocalizedMultilineTextInput extends React.Component {
  static displayName = 'LocalizedMultilineTextInput';

  static RequiredValueErrorMessage = RequiredValueErrorMessage;

  static propTypes = {
    id: PropTypes.string,
    name: PropTypes.string,
    // then input doesn't accept a "languages" prop, instead all possible
    // languages have to exist (with empty or filled strings) on the value:
    //   { en: 'foo', de: '', es: '' }
    value: PropTypes.objectOf(PropTypes.string).isRequired,
    onChange: requiredIf(PropTypes.func, props => !props.isReadOnly),
    selectedLanguage: PropTypes.string.isRequired,
    onBlur: PropTypes.func,
    onFocus: PropTypes.func,
    defaultExpandMultilineText: PropTypes.bool,
    hideLanguageExpansionControls: PropTypes.bool,
    defaultExpandLanguages: (props, propName, componentName, ...rest) => {
      if (
        props.hideLanguageExpansionControls &&
        typeof props[propName] === 'boolean'
      ) {
        throw new Error(
          oneLine`
            ${componentName}: "${propName}" does not have any effect when
            "hideLanguageExpansionControls" is set.
          `
        );
      }
      return PropTypes.bool(props, propName, componentName, ...rest);
    },
    isAutofocussed: PropTypes.bool,
    isDisabled: PropTypes.bool,
    isReadOnly: PropTypes.bool,
    placeholder: PropTypes.objectOf(PropTypes.string),
    horizontalConstraint: PropTypes.oneOf(['xs', 's', 'm', 'l', 'xl', 'scale']),
    hasError: PropTypes.bool,
    hasWarning: PropTypes.bool,
    errors: PropTypes.objectOf(PropTypes.node),
    warnings: PropTypes.objectOf(PropTypes.node),
    // HoC
    intl: PropTypes.shape({
      formatMessage: PropTypes.func.isRequired,
    }).isRequired,
  };

  static getId = getId;

  static getName = getName;

  static defaultProps = {
    horizontalConstraint: 'scale',
  };

  static createLocalizedString = createLocalizedString;

  static isEmpty = isEmpty;

  static omitEmptyTranslations = omitEmptyTranslations;

  static isTouched = isTouched;

  state = {
    // This state is used to show/hide the remaining translations
    areLanguagesOpened: this.props.defaultExpandLanguages,
    // This state is to manage the expand/collapse of multiline text inputs
    expandedTranslations: mapValues(this.props.value, () =>
      Boolean(this.props.defaultExpandMultilineText)
    ),
  };

  toggleLanguage = language =>
    this.setState(prevState => ({
      expandedTranslations: {
        ...prevState.expandedTranslations,
        [language]: !prevState.expandedTranslations[language],
      },
    }));

  toggleLanguages = () =>
    this.setState(prevState => ({
      areLanguagesOpened: !prevState.areLanguagesOpened,
    }));

  expandAllTranslations = () =>
    this.setState(prevState => ({
      expandedTranslations: mapValues(
        prevState.expandedTranslations,
        () => true
      ),
    }));

  static getDerivedStateFromProps = (props, state) => {
    // We want to automatically open the languages when an error is present on a
    // hidden input, and we want to keep the languages open even after the
    // error was resolved, so that the user can collapse it manually.
    // Otherwise it would close as soon as the error disappears.
    const hasErrorOnRemainingLanguages =
      props.hasError ||
      getHasErrorOnRemainingLanguages(props.errors, props.selectedLanguage);
    const hasWarningOnRemainingLanguages =
      props.hasWarning ||
      getHasWarningOnRemainingLanguages(props.warnings, props.selectedLanguage);
    const areLanguagesOpened =
      hasErrorOnRemainingLanguages ||
      hasWarningOnRemainingLanguages ||
      props.hideLanguageExpansionControls ||
      state.areLanguagesOpened;
    return { areLanguagesOpened };
  };

  render() {
    const languages = sortLanguages(
      this.props.selectedLanguage,
      Object.keys(this.props.value)
    );
    return (
      <Constraints.Horizontal constraint={this.props.horizontalConstraint}>
        <Spacings.Stack scale="s">
          {languages.map((language, index) => {
            const isFirstLanguage = index === 0;
            if (!isFirstLanguage && !this.state.areLanguagesOpened) return null;
            const isLastLanguage = index === languages.length - 1;
            const hasRemainingLanguages = languages.length > 1;
            const hasErrorOnRemainingLanguages =
              this.props.hasError ||
              getHasErrorOnRemainingLanguages(
                this.props.errors,
                this.props.selectedLanguage
              );
            const hasWarningOnRemainingLanguages =
              this.props.hasWarning ||
              getHasWarningOnRemainingLanguages(
                this.props.warnings,
                this.props.selectedLanguage
              );
            return (
              <TranslationInput
                key={language}
                id={LocalizedMultilineTextInput.getId(this.props.id, language)}
                name={LocalizedMultilineTextInput.getName(
                  this.props.name,
                  language
                )}
                value={this.props.value[language]}
                onChange={this.props.onChange}
                language={language}
                isCollapsed={!this.state.expandedTranslations[language]}
                onToggle={() => this.toggleLanguage(language)}
                placeholder={
                  this.props.placeholder
                    ? this.props.placeholder[language]
                    : undefined
                }
                onBlur={this.props.onBlur}
                onFocus={this.props.onFocus}
                isAutofocussed={index === 0 && this.props.isAutofocussed}
                isDisabled={this.props.isDisabled}
                isReadOnly={this.props.isReadOnly}
                languagesControl={(() => {
                  if (
                    !hasRemainingLanguages ||
                    this.props.hideLanguageExpansionControls
                  )
                    return null;
                  if (isFirstLanguage && !this.state.areLanguagesOpened)
                    return (
                      <LanguagesControl
                        isClosed={true}
                        onClick={() => {
                          // expand all multiline language inputs in case the
                          // first one was expanded when all languages
                          // are shown
                          if (
                            this.state.expandedTranslations[
                              this.props.selectedLanguage
                            ]
                          ) {
                            this.expandAllTranslations();
                          }
                          this.toggleLanguages();
                        }}
                        remainingLanguages={languages.length - 1}
                      />
                    );
                  if (isLastLanguage)
                    return (
                      <LanguagesControl
                        onClick={this.toggleLanguages}
                        remainingLanguages={languages.length - 1}
                        isDisabled={Boolean(
                          this.props.hasError ||
                            hasErrorOnRemainingLanguages ||
                            this.props.hasWarning ||
                            hasWarningOnRemainingLanguages
                        )}
                      />
                    );
                  return null;
                })()}
                hasError={Boolean(
                  this.props.hasError ||
                    (this.props.errors && this.props.errors[language])
                )}
                hasWarning={Boolean(
                  this.props.hasWarning ||
                    (this.props.warnings && this.props.warnings[language])
                )}
                intl={this.props.intl}
                warning={this.props.warnings && this.props.warnings[language]}
                error={this.props.errors && this.props.errors[language]}
                {...createLocalizedDataAttributes(this.props, language)}
              />
            );
          })}
        </Spacings.Stack>
      </Constraints.Horizontal>
    );
  }
}

const Wrapped = injectIntl(LocalizedMultilineTextInput);
Wrapped.isTouched = LocalizedMultilineTextInput.isTouched;
Wrapped.omitEmptyTranslations =
  LocalizedMultilineTextInput.omitEmptyTranslations;
Wrapped.isEmpty = LocalizedMultilineTextInput.isEmpty;
Wrapped.createLocalizedString =
  LocalizedMultilineTextInput.createLocalizedString;
export default Wrapped;
