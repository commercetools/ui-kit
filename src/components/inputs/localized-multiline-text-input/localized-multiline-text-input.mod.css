:root {
  /* we need this line-height to achieve 32px height when the component has only one row */
  --size-input-line-height: 22px;
}

/* Base */

.fieldContainer {
  width: 100%;
  position: relative;
  display: flex;
}

.languageLabel {
  /* avoid wrapping label onto new lines */
  flex: 1 0 auto;
  color: var(--font-color-disabled);
  line-height: calc(
    var(--size-height-input) - 2 * var(--border-radius-1)
  );
  background-color: var(--background-color-input-disabled);
  border-top-left-radius: var(--border-radius-input);
  border-bottom-left-radius: var(--border-radius-input);
  border: 1px var(--border-color-input-disabled) solid;
  padding: 0 var(--spacing-8);
  transition: var(--transition-standard);
  border-right: 0;
  box-shadow: none;
  appearance: none;
}

.input {
  color: var(--font-color-default);
  font-family: var(--font-family-default);
  font-size: var(--font-size-default);
  min-height: var(--size-height-input);
  border-top-right-radius: var(--border-radius-input);
  border-bottom-right-radius: var(--border-radius-input);
  width: 100%;
  border-style: solid;
  border-width: var(--border-radius-1);
  padding: var(--spacing-4) var(--spacing-8);
  transition: var(--transition-standard);
  outline: 0;
  box-shadow: none;
  appearance: none;
  line-height: var(--size-input-line-height);
  white-space: pre-wrap;
  resize: vertical;
}

.inputClosed {
  overflow: hidden;
  word-break: break-all;
}

.input::placeholder {
  color: var(--font-color-placeholder);
}

/* States */

.pristine {
  composes: input;
  border-color: var(--border-color-input-pristine);
  background-color: var(--background-color-input-pristine);
}

.warning {
  composes: input;
  color: var(--font-color-warning);
  border-color: var(--border-color-input-warning);
}

.error {
  composes: input;
  color: var(--font-color-error);
  border-color: var(--border-color-input-error);
}

.disabled {
  composes: input;
  cursor: not-allowed;
  color: var(--font-color-disabled);
  border-color: var(--border-color-input-disabled);
  background-color: var(--background-color-input-disabled);
  opacity: 1; /* fix for mobile safari */
}

.readonly {
  composes: input;
  border-color: var(--border-color-input-readonly);
  color: var(--font-color-readonly);
  cursor: default;
}

/* States */

.pristine:active,
.pristine:focus {
  border-color: var(--border-color-input-focus);
  background: var(--background-color-input-pristine);
  color: var(--font-color-default);
}

/* Commands */

.commandsContainer {
  display: flex;
}

.commandsLeft {
  flex: 1;
}

.commandsExpand {
  flex: 0;
}
