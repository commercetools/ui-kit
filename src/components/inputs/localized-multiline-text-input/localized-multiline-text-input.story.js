import React from 'react';
import { Value } from 'react-value';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import {
  withKnobs,
  boolean,
  text,
  select,
  object,
} from '@storybook/addon-knobs/react';
import Section from '../../../../.storybook/decorators/section';
import ErrorMessage from '../../messages/error-message';
import WarningMessage from '../../messages/warning-message';
import Readme from './README.md';
import LocalizedMultilineTextInput from './localized-multiline-text-input';

storiesOf('Components|Inputs', module)
  .addDecorator(withKnobs)
  .addParameters({
    readme: {
      // Show readme at the addons panel
      sidebar: Readme,
    },
  })
  .add('LocalizedMultilineTextInput', () => {
    const defaultExpandLanguages = boolean('defaultExpandLanguages', false);
    const defaultExpandMultilineText = boolean(
      'defaultExpandMultilineText',
      false
    );
    const errors = object('errors', { en: '', de: '', 'nan-Hant-TW': '' });
    const warnings = object('warnings', { en: '', de: '', 'nan-Hant-TW': '' });
    // We need to force the component to rerender in case a default value
    // is changed. Otherwise the knob would have no effect.
    // We do this by changing the key.
    const key = `${defaultExpandMultilineText}-${defaultExpandLanguages}`;
    return (
      <Section>
        <Value
          defaultValue={{ en: '', de: '', 'nan-Hant-TW': '' }}
          render={(value, onChange) => (
            <LocalizedMultilineTextInput
              key={key}
              id={text('id', undefined)}
              name={text('name', 'productName')}
              value={object('value', value)}
              onChange={event => {
                action('onChange')(event);
                onChange({
                  ...value,
                  [event.target.language]: event.target.value,
                });
              }}
              selectedLanguage={select('selectedLanguage', ['en', 'de'], 'en')}
              onBlur={action('onBlur')}
              onFocus={action('onFocus')}
              hideLanguageExpansionControls={boolean(
                'hideLanguageExpansionControls',
                false
              )}
              defaultExpandLanguages={
                // we need to set undefined instead of false to avoid prop-type
                // warnings in case hideLanguageExpansionControls is true
                defaultExpandLanguages || undefined
              }
              defaultExpandMultilineText={defaultExpandMultilineText}
              isAutofocussed={boolean('isAutofocussed', false)}
              isDisabled={boolean('isDisabled', false)}
              isReadOnly={boolean('isReadOnly', false)}
              placeholder={object('placeholder', { en: '', de: '' })}
              horizontalConstraint={select(
                'horizontalConstraint',
                ['m', 'l', 'xl', 'scale'],
                'm'
              )}
              hasError={boolean('hasError', false)}
              hasWarning={boolean('hasWarning', false)}
              errors={
                Object.values(errors).some(error => error.length > 0)
                  ? Object.entries(errors).reduce((acc, [language, error]) => {
                      if (error.length === 0) return acc;
                      acc[language] = <ErrorMessage>{error}</ErrorMessage>;
                      return acc;
                    }, {})
                  : undefined
              }
              warnings={
                Object.values(warnings).some(warning => warning.length > 0)
                  ? Object.entries(warnings).reduce(
                      (acc, [language, warning]) => {
                        if (warning.length === 0) return acc;
                        acc[language] = (
                          <WarningMessage>{warning}</WarningMessage>
                        );
                        return acc;
                      },
                      {}
                    )
                  : undefined
              }
              data-test="foo"
            />
          )}
        />
      </Section>
    );
  });
