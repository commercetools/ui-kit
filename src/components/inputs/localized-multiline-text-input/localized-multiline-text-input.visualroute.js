import React from 'react';
import {
  LocalizedMultilineTextInput,
  ErrorMessage,
  WarningMessage,
} from 'ui-kit';
import { Suite, Spec } from '../../../../test/percy';

const value = {
  en: 'hello\nworld',
  de: 'hallo\nwelt',
  es: 'hola\nmundo',
};

export const routePath = '/localized-multiline-text-input';

export const component = () => (
  <Suite>
    <Spec label="minimal">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
      />
    </Spec>
    <Spec label="when multiline text is expanded by default">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        isMultilineDefaultExpanded={true}
      />
    </Spec>
    <Spec label="when multiline text and languages are expanded by default">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        isMultilineDefaultExpanded={true}
        areLanguagesDefaultOpened={true}
      />
    </Spec>
    <Spec label="when language controls are hidden">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        hideLanguageControls={true}
      />
    </Spec>
    <Spec label="when languages are opened by default">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        areLanguagesDefaultOpened={true}
      />
    </Spec>
    <Spec label="when read-only and open">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        isReadOnly={true}
        areLanguagesDefaultOpened={true}
      />
    </Spec>
    <Spec label="when read-only and closed">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        isReadOnly={true}
      />
    </Spec>
    <Spec label="when disabled and open">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        isDisabled={true}
        areLanguagesDefaultOpened={true}
      />
    </Spec>
    <Spec label="when disabled and closed">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        isDisabled={true}
      />
    </Spec>
    <Spec label="when there is an error for a specific language (first one)">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        errors={{ en: <ErrorMessage>foo</ErrorMessage> }}
      />
    </Spec>
    <Spec label="when there is an error for a specific language (second one)">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        errors={{ de: <ErrorMessage>foo</ErrorMessage> }}
      />
    </Spec>
    <Spec label="when there is a general error">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        hasError={true}
      />
    </Spec>
    <Spec label="when there is a warning for a specific language (first one)">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        warnings={{ en: <WarningMessage>foo</WarningMessage> }}
      />
    </Spec>
    <Spec label="when there is a warning for a specific language (second one)">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        warnings={{ de: <WarningMessage>foo</WarningMessage> }}
      />
    </Spec>
    <Spec label="when there is a general warning">
      <LocalizedMultilineTextInput
        value={value}
        onChange={() => {}}
        selectedLanguage="en"
        horizontalConstraint="m"
        hasWarning={true}
      />
    </Spec>
  </Suite>
);
