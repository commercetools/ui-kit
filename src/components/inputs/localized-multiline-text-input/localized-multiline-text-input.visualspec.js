import { percySnapshot } from '@percy/puppeteer';

describe('LocalizedMultilineTextInput', () => {
  beforeAll(async () => {
    await page.goto(`${HOST}/localized-multiline-text-input`);
  });

  it('Default', async () => {
    await expect(page).toMatch('minimal');
    await percySnapshot(page, 'LocalizedMultilineTextInput');
  });
});
