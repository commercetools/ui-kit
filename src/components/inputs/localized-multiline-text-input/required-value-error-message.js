import React from 'react';
import { FormattedMessage } from 'react-intl';
import ErrorMessage from '../../messages/error-message';
import messages from '../../internals/messages/localized-input';

const RequiredValueErrorMessage = () => (
  <ErrorMessage>
    <FormattedMessage {...messages.missingRequiredField} />
  </ErrorMessage>
);

RequiredValueErrorMessage.displayName = 'RequiredValueErrorMessage';

export default RequiredValueErrorMessage;
