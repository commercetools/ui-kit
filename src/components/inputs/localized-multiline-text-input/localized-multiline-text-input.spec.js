import React from 'react';
import PropTypes from 'prop-types';
import { render, fireEvent } from '../../../test-utils';
import { LocalizedMultilineTextInput } from './localized-multiline-text-input';

// We use this component to simulate the whole flow of
// changing a value and formatting on blur.
class TestComponent extends React.Component {
  static displayName = 'TestComponent';
  static propTypes = {
    id: PropTypes.string,
    value: PropTypes.objectOf(PropTypes.string).isRequired,
    onChange: PropTypes.func,
    selectedLanguage: PropTypes.string.isRequired,
  };
  static defaultProps = {
    id: 'some-id',
    name: 'some-name',
    value: {
      en: 'Oh no, oh no, oh no. no',
      fr: "Calisse d'ostie de tabarnak",
    },
    selectedLanguage: 'en',
    intl: { formatMessage: message => message.id },
  };

  state = {
    value: this.props.value || {},
    selectedLanguage: this.props.selectedLanguage || '',
  };

  handleChange = event => {
    event.persist();
    this.setState({
      value: {
        ...this.state.value,
        [event.target.language]: event.target.value,
      },
    });
  };

  render() {
    return (
      <LocalizedMultilineTextInput
        {...this.props}
        onChange={this.props.onChange || this.handleChange}
        value={this.state.value}
        selectedLanguage={this.state.selectedLanguage}
      />
    );
  }
}

const renderLocalizedMultilineTextInput = (props, options) =>
  render(<TestComponent {...props} />, options);

it('should forward data-attributes', () => {
  const { container } = renderLocalizedMultilineTextInput({
    'data-foo': 'bar',
  });
  expect(container.querySelector('[data-foo="bar"]')).toBeInTheDocument();
});

it('should have an HTML name', () => {
  const { getByLabelText } = renderLocalizedMultilineTextInput({
    name: 'foo',
  });
  expect(getByLabelText('EN')).toHaveAttribute('name', 'foo.en');
});

it('should have focus automatically when isAutofocussed is passed', () => {
  const { getByLabelText } = renderLocalizedMultilineTextInput({
    isAutofocussed: true,
  });
  expect(getByLabelText('EN')).toHaveFocus();
});

describe('when collapsed', () => {
  it('should render an input for the selected language (en)', () => {
    const { getByLabelText } = renderLocalizedMultilineTextInput();
    expect(getByLabelText('EN')).toBeInTheDocument();
  });
});

it('should call onFocus when the input is focused', () => {
  const onFocus = jest.fn();
  const { getByLabelText } = renderLocalizedMultilineTextInput({ onFocus });
  const input = getByLabelText('EN');
  input.focus();
  expect(input).toHaveFocus();
  expect(onFocus).toHaveBeenCalled();
});

it('should call onBlur when input loses focus', () => {
  const onBlur = jest.fn();
  const { getByLabelText } = renderLocalizedMultilineTextInput({ onBlur });
  const input = getByLabelText('EN');
  input.focus();
  expect(input).toHaveFocus();
  input.blur();
  expect(input).not.toHaveFocus();
  expect(onBlur).toHaveBeenCalled();
});

describe('when input is collapsed', () => {
  it('should allow changing value of the input', () => {
    const { getByLabelText } = renderLocalizedMultilineTextInput();
    const event = { target: { value: 'I want chicken' } };
    const input = getByLabelText('EN');
    fireEvent.focus(input);
    fireEvent.change(input, event);
    fireEvent.keyDown(input, { key: 'Enter' });
    fireEvent.keyUp(input, { key: 'Enter' });
    expect(input.value).toEqual('I want chicken');
  });
});

describe('when input is expanded', () => {
  it('should expand and show all language inputs when `Show all languages` is clicked', () => {
    const { getByLabelText } = renderLocalizedMultilineTextInput();
    getByLabelText(/show all languages/i).click();
    expect(getByLabelText('FR')).toBeInTheDocument();
  });
  it('should allow changing the french input', () => {
    const { getByLabelText } = renderLocalizedMultilineTextInput();
    getByLabelText(/show all languages/i).click();
    const event = { target: { value: 'Je veux manger du poulet' } };
    const frenchInput = getByLabelText('FR');
    fireEvent.focus(frenchInput);
    fireEvent.change(frenchInput, event);
    fireEvent.keyDown(frenchInput, { key: 'Enter' });
    fireEvent.keyDown(frenchInput, { key: 'Enter' });
    expect(frenchInput.value).toEqual('Je veux manger du poulet');
  });
});

describe('when expanded by default', () => {
  it('should render one input per language', () => {
    const { getByLabelText } = renderLocalizedMultilineTextInput({
      areLanguagesDefaultOpened: true,
    });
    expect(getByLabelText('EN')).toBeInTheDocument();
    expect(getByLabelText('FR')).toBeInTheDocument();
    expect(getByLabelText(/hide languages/i)).toBeInTheDocument();
  });
});

describe('when expansion controls are hidden', () => {
  it('should render one input per language and no hide button', () => {
    const {
      getByLabelText,
      queryByLabelText,
    } = renderLocalizedMultilineTextInput({
      hideLanguageControls: true,
    });
    expect(getByLabelText('EN')).toBeInTheDocument();
    expect(getByLabelText('FR')).toBeInTheDocument();
    expect(queryByLabelText(/hide languages/i)).not.toBeInTheDocument();
  });
});

describe('when disabled', () => {
  describe('when not expanded', () => {
    it('should render a disabled input', () => {
      const { getByLabelText } = renderLocalizedMultilineTextInput({
        isDisabled: true,
      });
      expect(getByLabelText('EN')).toHaveAttribute('disabled');
    });
  });
  describe('when expanded', () => {
    it('should be able to expand, and all inputs are disabled', () => {
      const { getByLabelText } = renderLocalizedMultilineTextInput({
        isDisabled: true,
      });
      getByLabelText(/show all languages/i).click();
      expect(getByLabelText('EN')).toHaveAttribute('disabled');
      expect(getByLabelText('FR')).toHaveAttribute('disabled');
    });
  });
});

describe('when read-only', () => {
  describe('when not expanded', () => {
    it('should render a readonly input', () => {
      const {
        getByLabelText,
        queryByLabelText,
      } = renderLocalizedMultilineTextInput({
        isReadOnly: true,
      });
      expect(getByLabelText('EN')).toHaveAttribute('readonly');
      expect(queryByLabelText('FR')).not.toBeInTheDocument();
    });
  });
  describe('when expanded', () => {
    it('should be able to expand, and all inputs are readonly', () => {
      const { getByLabelText } = renderLocalizedMultilineTextInput({
        isReadOnly: true,
      });
      getByLabelText(/show all languages/i).click();
      expect(getByLabelText('EN')).toHaveAttribute('readonly');
      expect(getByLabelText('FR')).toHaveAttribute('readonly');
    });
  });
});

describe('when placeholders are provided', () => {
  it('should forward the placeholders', () => {
    const { getByLabelText } = renderLocalizedMultilineTextInput({
      areLanguagesDefaultOpened: true,
      placeholder: {
        en: 'I love cheese',
        fr: "Ostie que j'aime le fromage",
      },
    });
    expect(getByLabelText('EN')).toHaveAttribute(
      'placeholder',
      'I love cheese'
    );
    expect(getByLabelText('FR')).toHaveAttribute(
      'placeholder',
      "Ostie que j'aime le fromage"
    );
  });
});

describe('when every field should display an error', () => {
  const errors = {
    en: 'A value is required',
    fr: "Ostie, t'as oublier de manger du fromage",
  };
  it('should be open all fields and render errors', () => {
    const { getByLabelText, getByText } = renderLocalizedMultilineTextInput({
      errors,
    });
    expect(getByLabelText('EN')).toBeInTheDocument();
    expect(getByLabelText('FR')).toBeInTheDocument();
    expect(getByText(errors.en)).toBeInTheDocument();
    expect(getByText(errors.fr)).toBeInTheDocument();
  });
});

describe('when the error is not on the selected language', () => {
  const errors = {
    en: '',
    fr: "Ostie, t'as oublier de manger du fromage",
  };
  it('should be open all fields and render errors', () => {
    const { getByLabelText, getByText } = renderLocalizedMultilineTextInput({
      errors,
    });
    expect(getByLabelText('EN')).toBeInTheDocument();
    expect(getByLabelText('FR')).toBeInTheDocument();
    expect(getByText(errors.fr)).toBeInTheDocument();
  });
});

describe('when the error is on the selected language', () => {
  it('should display the error without expanding', () => {
    const {
      getByLabelText,
      getByText,
      queryByLabelText,
    } = renderLocalizedMultilineTextInput({
      errors: {
        en: 'Some error',
      },
    });
    expect(getByLabelText('EN')).toBeInTheDocument();
    expect(queryByLabelText('FR')).not.toBeInTheDocument();
    expect(getByText('Some error')).toBeInTheDocument();
  });
});
