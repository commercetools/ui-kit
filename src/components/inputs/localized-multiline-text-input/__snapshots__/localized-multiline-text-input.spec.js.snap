// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering other languages should sort them by the relevance 1`] = `
<Horizontal
  constraint="scale"
>
  <Stack
    alignItems="stretch"
    scale="s"
  >
    <TranslationInput
      hasError={false}
      hasWarning={false}
      intl={
        Object {
          "formatMessage": [Function],
        }
      }
      isCollapsed={true}
      key="en"
      language="en"
      languagesControl={null}
      onChange={[MockFunction]}
      onToggle={[Function]}
      value="en"
    />
    <TranslationInput
      hasError={false}
      hasWarning={false}
      intl={
        Object {
          "formatMessage": [Function],
        }
      }
      isAutofocussed={false}
      isCollapsed={true}
      key="en-GB"
      language="en-GB"
      languagesControl={null}
      onChange={[MockFunction]}
      onToggle={[Function]}
      value="en-GB"
    />
    <TranslationInput
      hasError={false}
      hasWarning={false}
      intl={
        Object {
          "formatMessage": [Function],
        }
      }
      isAutofocussed={false}
      isCollapsed={true}
      key="de"
      language="de"
      languagesControl={null}
      onChange={[MockFunction]}
      onToggle={[Function]}
      value="de"
    />
    <TranslationInput
      hasError={false}
      hasWarning={false}
      intl={
        Object {
          "formatMessage": [Function],
        }
      }
      isAutofocussed={false}
      isCollapsed={true}
      key="fr"
      language="fr"
      languagesControl={null}
      onChange={[MockFunction]}
      onToggle={[Function]}
      value="fr"
    />
    <TranslationInput
      hasError={false}
      hasWarning={false}
      intl={
        Object {
          "formatMessage": [Function],
        }
      }
      isAutofocussed={false}
      isCollapsed={true}
      key="nan-Hant-TW"
      language="nan-Hant-TW"
      languagesControl={null}
      onChange={[MockFunction]}
      onToggle={[Function]}
      value="nan-Hant-TW"
    />
    <TranslationInput
      hasError={false}
      hasWarning={false}
      intl={
        Object {
          "formatMessage": [Function],
        }
      }
      isAutofocussed={false}
      isCollapsed={true}
      key="pt-BR"
      language="pt-BR"
      languagesControl={
        <InjectIntl(LanguagesControl)
          isDisabled={false}
          onClick={[Function]}
          remainingLanguages={5}
        />
      }
      onChange={[MockFunction]}
      onToggle={[Function]}
      value="pt-BR"
    />
  </Stack>
</Horizontal>
`;

exports[`rendering with minimal props should match snapshot 1`] = `
<Horizontal
  constraint="scale"
>
  <Stack
    alignItems="stretch"
    scale="s"
  >
    <TranslationInput
      hasError={false}
      hasWarning={false}
      intl={
        Object {
          "formatMessage": [Function],
        }
      }
      isCollapsed={true}
      key="en"
      language="en"
      languagesControl={
        <InjectIntl(LanguagesControl)
          isClosed={true}
          onClick={[Function]}
          remainingLanguages={1}
        />
      }
      onChange={[MockFunction]}
      onToggle={[Function]}
      value="Horse"
    />
  </Stack>
</Horizontal>
`;
